File: 01. [Part 1]- Java & Kotlin/38. Architecture Patterns MVC MVP MVVM/03. MVVM Pattern View Binding/app/src/main/java/com/mastercoding/patternsApp/ui/MainActivity.java
 package com.mastercoding.patternsApp.ui;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;
import androidx.lifecycle.ViewModelProviders;

import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

import com.mastercoding.patternsApp.R;
import com.mastercoding.patternsApp.util.MyModel;

 public class MainActivity extends AppCompatActivity{

    Button btn;
    TextView textView;

    // 6- Connecting MainActivity with ViewModel and Listening to LIVE Data
     AppViewModel appViewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        btn = findViewById(R.id.button);
        textView = findViewById(R.id.textView);


        btn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                    appViewModel.getAppName();
            }
        });

        // instead of using deprecated methods
        appViewModel = new ViewModelProvider(this).get(AppViewModel.class);

        // 7- Listening & observing the changes to Live Data
        appViewModel.mutableLiveData.observe(this, new Observer<String>() {
            @Override
            public void onChanged(String s) {
                 // Any Change in the live data do this:
                textView.setText(s);

            }
        });


    }

     //     MVC Pattern
    // View: TextView & Button (Main Activity)
    // Model: MyModel class
    // Controller: MainActivity


     //      MVP Pattern
     // View : Activity with Textview & Button
     // Model: My Model
     // Presenter: AppPresenter


     // MVVM
     // 1- Implementing the dependency of ViewModel
     // 2- Creating ViewModel Class

     //Amazing Guys !! We have implemented MVVM correctly
}

File: 01. [Part 1]- Java & Kotlin/41. Kotlin Mastering Retrofit Top Companies App/app/src/main/java/com/mastercoding/retrofitdemoapp/MainActivity.kt
package com.mastercoding.retrofitdemoapp

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.widget.TextView
import androidx.lifecycle.LiveData
import androidx.lifecycle.Observer
import androidx.lifecycle.liveData
import retrofit2.Response

class MainActivity : AppCompatActivity() {
    lateinit var resultTextView: TextView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        resultTextView = findViewById(R.id.textView2)

        // Retrofit Instance
        val retrofitInstance = RetrofitInstance
            .getRetrofitInstance()
            .create(CompaniesService::class.java)

        // Live Data : Fetch Data Asynchronously
        val responseLiveData : LiveData<Response<CompaniesList>> =
            liveData {
                // Asynchronous operation is defined here

                val countriesResponse = retrofitInstance.getAllCompanies()
                emit(countriesResponse)

            }

        // Observing data changes in LiveData
        responseLiveData.observe(this, Observer {

            // execute this code when the LiveData's value changes
            val companiesList = it.body()?.listIterator()

            if (companiesList != null){
                while (companiesList.hasNext()){
                    val companyItem = companiesList.next()

                    val companiesResult = " Company Name: ${companyItem.name} \n"
                    resultTextView.append(companiesResult)
                }
            }

        })


    }
}